### DISCUSSION TITLE
constant Option is not recognised as constant value in inline if

### DISCUSSION THREAD

**Discussion by rssh** [scala/scala3#17789] [Feature Requests]
## Compiler version

current (Tue Jan 11 08:53:51 EET 2022) master

## Minimized code

```Scala

object Test {

  inline def get: Option[String] = Some("")

  inline def fun: Unit =
    inline if (get.isDefined) then
      println("value = ${get.get}")
    else
      println("value not set")

  fun

}

```

## Output

```scala
[info] welcome to sbt 1.5.5 (Homebrew Java 17)
[info] loading global plugins from /Users/rssh/.sbt/1.0/plugins
[info] loading project definition from /Users/rssh/tests/dotty/compileTimeIfOpt/project
[info] loading settings for project root from build.sbt ...
[info] set current project to test (in build file:/Users/rssh/tests/dotty/compileTimeIfOpt/)
[info] Executing in batch mode. For better performance use sbt's shell
[info] compiling 1 Scala source to /Users/rssh/tests/dotty/compileTimeIfOpt/target/scala-3.1.2-RC1-bin-SNAPSHOT/classes ...
[error] -- Error: /Users/rssh/tests/dotty/compileTimeIfOpt/src/main/scala/x/Test.scala:13:2 
[error] 13 |  fun
[error]    |  ^^^
[error]    |Cannot reduce `inline if` because its condition is not a constant value: (Some.apply[String](""):Option[String]).isDefined
[error]    | This location contains code that was inlined from Test.scala:8
```

## Expectation

 Should be compiled.

**Comment by rssh** [comment]
inverse to lampepfl/dotty#14245

**Comment by rssh** [comment]
In case of .isDefined is problematic:

```
object Test {

  inline def get: Option[String] = Some("")

  inline def fun: Unit =
    inline if (get != None) then
      println("value = ${get.get}")
    else
      println("value not set")

  fun

}
```

result:
```
[error] -- Error: /Users/rssh/tests/dotty/compileTimeIfOpt/src/main/scala/x/Test.scala:13:2 
[error] 13 |  fun
[error]    |  ^^^
[error]    |Cannot reduce `inline if` because its condition is not a constant value: (Some.apply[String](""):Option[String]).!=(None)
```

**Comment by rssh** [comment]
Not sure, this is a bug or not ?

**Comment by dwijnand** [comment]
Doesn't look like a bug to me.  There's no way the compiler can tell that `(Some.apply[String](""):Option[String]).!=(None)`.  Even `new Some("") != None` isn't certain, because equals could be defined to be true there.

**Comment by rssh** [comment]
I'm not against closing this issue,  but want to leave a remark just from the loving to argue: 

`Option` has implemented `FromExpr`,  so a compiler can call `FromExpr.unapply` to receive an object.  And  == and !=. are parsed specially,  because `inline if (1 == 2)`. works.  So,  it is possible to check the existence of FromExpr in '=='.   

On another side, you are right, that this is more a feature request than a bug.

**Comment by nicolasstucki** [comment]
Any kind of evaluation performed by inlining must be hardcoded in the compiler. The more we make it do, the slower it will become. 

On the other hand, with macros, the code is in the library (such as `FromExpr `) which can contain arbitrarily complex partial evaluation abstractions which scale with the use cases. These also are executing compiled code which is faster than full AST interpretation.

**Comment by dwijnand** [comment]
> `Option` has implemented `FromExpr`, so a compiler can call `FromExpr.unapply` to receive an object. And == and !=. are parsed specially, because `inline if (1 == 2)`. works. So, it is possible to check the existence of FromExpr in '=='.
> 
> On another side, you are right, that this is more a feature request than a bug.

TIL FromExpr! 😄   Sounds like a plausible feature request to me, so feel free to rehash this as a request in lampepfl/dotty-feature-requests.
